toolchain/parse/handle_declaration_scope_loop.cpp

```diff
@@ -22,75 +22,106 @@ static auto HandleUnrecognizedDecl(Context& context) -> void {
 auto HandleDeclScopeLoop(Context& context) -> void {
   // This maintains the current state unless we're at the end of the scope.
 
-  switch (auto position_kind = context.PositionKind()) {
+  auto position_kind = context.PositionKind();
+  switch (position_kind) {
     case Lex::TokenKind::CloseCurlyBrace:
     case Lex::TokenKind::EndOfFile: {
       // This is the end of the scope, so the loop state ends.
       context.PopAndDiscardState();
-      break;
+      return;
     }
     // `import` and `package` manage their packaging state.
     case Lex::TokenKind::Import: {
       context.PushState(State::Import);
-      break;
+      return;
     }
     case Lex::TokenKind::Package: {
       context.PushState(State::Package);
+      return;
+    }
+    default: {
       break;
     }
-    default:
-      // Because a non-packaging keyword was encountered, packaging is complete.
-      // Misplaced packaging keywords may lead to this being re-triggered.
-      if (context.packaging_state() !=
-          Context::PackagingState::AfterNonPackagingDecl) {
-        if (!context.first_non_packaging_token().is_valid()) {
-          context.set_first_non_packaging_token(*context.position());
-        }
-        context.set_packaging_state(
-            Context::PackagingState::AfterNonPackagingDecl);
+  }
+
+  // Because a non-packaging keyword was encountered, packaging is complete.
+  // Misplaced packaging keywords may lead to this being re-triggered.
+  if (context.packaging_state() !=
+      Context::PackagingState::AfterNonPackagingDecl) {
+    if (!context.first_non_packaging_token().is_valid()) {
+      context.set_first_non_packaging_token(*context.position());
+    }
+    context.set_packaging_state(Context::PackagingState::AfterNonPackagingDecl);
+  }
+
+  // Remaining keywords are only valid after imports are complete, and so all
+  // result in a `set_packaging_state` call. Note, this may not always be
+  // necessary but is probably cheaper than validating.
+  switch (position_kind) {
+    case Lex::TokenKind::Abstract: {
```

Comment:
What would you think of pushing `abstract` and `base` handling out of this file, and into their own state? That way, the general declaration scope handling doesn't start to become specialized towards handling specific syntaxes, which would make the file scale up in size. Perhaps that would also allow for aa setup that didn't need to change the handling of `interface` or `class` in this file, allowing `Consume` to continue to be done locally within handlers.

Reply:
I'm a little wary here, particularly due to the addition of peeking in PositionIs and PositionKind to support something that I think should be removed. We can proceed, but can you add a TODO to those functions to suggest avoiding/removing the depth argument if possible?